gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSub.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import com.google.cloud.AsyncPage;
+import com.google.cloud.Page;
+import com.google.cloud.Service;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An interface for Google Cloud Pub/Sub.
+ *
+ * @see <a href="https://cloud.google.com/pubsub/">Google Cloud Pub/Sub</a>
+ */
+public interface PubSub extends Service<PubSubOptions> {
+
+  final class ListOption implements Serializable {
+
+    private static final long serialVersionUID = 6517442127283383124L;
+
+    private final Option option;
+    private final Object value;
+
+    enum Option {
+      PAGE_SIZE, PAGE_TOKEN
+    }
+
+    private ListOption(Option option, Object value) {
+      this.option = option;
+      this.value = value;
+    }
+
+    Option option() {
+      return option;
+    }
+
+    Object value() {
+      return value;
+    }
+
+    public static ListOption pageSize(int pageSize) {
+      return new ListOption(Option.PAGE_SIZE, pageSize);
+    }
+
+    public static ListOption pageToken(String pageToken) {
+      return new ListOption(Option.PAGE_TOKEN, pageToken);
+    }
+  }
+
+  final class PullOption implements Serializable {
+
+    private static final long serialVersionUID = -5220474819637439937L;
+
+    private final Option option;
+    private final Object value;
+
+    enum Option {
+      RETURN_IMMEDIATELY, MAX_MESSAGES
+    }
+
+    private PullOption(Option option, Object value) {
+      this.option = option;
+      this.value = value;
+    }
+
+    Option option() {
+      return option;
+    }
+
+    Object value() {
+      return value;
+    }
+
+    public static PullOption returnImmediatly() {
+      return new PullOption(Option.RETURN_IMMEDIATELY, true);
+    }
+
+    public static PullOption maxMessages(int maxMessages) {
+      return new PullOption(Option.MAX_MESSAGES, maxMessages);
+    }
+  }
+
+  /**
+   * A callback to process pulled messages.
+   * The message will be ack'ed upon successful return or nack'ed if exception is thrown.
+   */
+  interface PullCallback {
+    void process(Message message) throws Exception;
+  }
+
+  Topic create(TopicInfo topic);
```

Comment:
I'm completely fine if we just leave the VTK API; and have this library just be a class that is a helper to manage the PullConsumer.

I really don't see the value in wrapping every method and data type with equivalent classes that are mostly boilerplate. If that is necessary, it would mean that the VTK project is not doing its job correctly, which is IIUC to wrap the gRPC generated client libraries with more idiomatic syntax. I do see small improvements on the wrappers; but overall not enough to justify them.

I also thought that VTK libraries would allow to extend a service by adding custom methods; but that may have been an old idea.